### PR TITLE
fix(eslint-plugin): [strict-boolean-expression] fix nullable object handling in LogicalExpression

### LINKED ISSUE #7743: Bug: [strict-boolean-expressions] autofix can be incorrect with allowNullableObject set to false
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.2.2&fileType=.tsx&code=LAKAxg9gdgzgLgAgGYQgIwIYCcEF4EYwCeUYCAFEgK6kD8AXBQJR4B8CUVANly7uwG9QCBJFiIkASygYuAMRplcwkckUIAPhpUjy5Pu048mAbhVYApnCpYoyabIWl9ZkAF8TQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYkACAXAngA4oDG0AlobgLQrzkB2uA9MrhSTQEYD2PSAQwa0AHoWgpk5Hg1QYA2pETRoPaJAA0mcBCgD48HgHcAyu0YBzdGABm%2B5Ig06IkfYaMA5WAFsuy63bwDk66egbGXgYCXEgA8lwAVoicAfaOzmHukfDRSABCfIIMqUHpoa7hngg5MYhmFAxWGIHBGRVZ1bmIXr7%2BzWkhuu0RnbUAogw%2BJa3lbsYAggz402VDc0YASgiIACo8WwwA6uS4ABY8sLj15JzZAMKnyQDWyACSANIMxoenAriv828ABEeJYVhkIABfAC6kJAkKAA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```typescript
const foobar = async (func?: () => null) => {
  const finalFunc =
    func ||
    (() => null);
  return finalFunc();
};
```

### ESLint Config

```javascript
module.exports = {
  "rules": {"@typescript-eslint/strict-boolean-expressions": ["error", {
    "allowString": false,
    "allowNumber": false,
    "allowNullableObject": false,
    "allowNullableBoolean": false,
    "allowNullableString": false,
    "allowNullableNumber": false,
    "allowNullableEnum": false,
    "allowAny": false,
    "allowRuleToRunWithoutStrictNullChecksIKnowWhatIAmDoing": false
      }]}
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

The auto fix introduces a bug in this case, as it replaces `func || (() => null);` with `func != null  || (() => null);` which is now either a boolean or a function, when the original code was such that it always resulted in a function.

In this case there should be no auto fix, or the actual fix is to use nullish coalescing (`func ?? (() => null);`), but that seems a bit beyond the scope of this rule...

### Actual Result

The auto fix introduces a bug in this case, as it replaces `func || (() => null);` with `func != null  || (() => null);` which is now either a boolean or a function, when the original code was such that it always resulted in a function.

### Additional Info

_No response_

### PR DESCRIPTION
**Pull request by barroudjo**
## PR Checklist

- [x] Addresses an existing open issue: fixes #7743
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

When a nullable object appears in a logical expression other than nullish coalescing (so `||` or `&&`), the suggested fix could be wrong and introduce a bug (as explained in the [issue](https://github.com/typescript-eslint/typescript-eslint/issues/7743)). This MR removes the suggested fix for these cases, and gives a more explicit message.

### CODE REVIEW

File: packages/eslint-plugin/src/rules/strict-boolean-expressions.ts [hunk 1]
```diff
@@ -937,6 +947,14 @@ function isLogicalNegationExpression(
   return node.type === AST_NODE_TYPES.UnaryExpression && node.operator === '!';
 }
 
+function isLogicalExpressionNotNullishCoalescing(
+  node: TSESTree.Node,
+): node is TSESTree.LogicalExpression {
+  return (
+    node.type === AST_NODE_TYPES.LogicalExpression && node.operator !== '??'
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/strict-boolean-expressions.ts:954]: > `&& node.operator !== '??'`

I tried removing this added check locally and no unit tests failed. Either the check is unnecessary or there's a bag in unit test coverage. My hunch is the latter, but haven't looked too deeply. Requesting changes either way.

**Review by JoshuaKGoldberg — APPROVED** [review]
Looks great, very close! Just requesting changes on the lack of coverage. Thanks! ✨ 

Great first contribution 🙂

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
_(wrong button first)_

**Comment by JoshuaKGoldberg** [comment]
👋 Hey @barroudjo! Just checking in, is this still something you have time for? No worries if not - I just don't want to leave it hanging.

**Comment by barroudjo** [comment]
You're right, I need to get back to it. I should have the time next week.

**Comment by JoshuaKGoldberg** [comment]
Closing this PR as it's been stale for a while without activity. cc @barroudjo - if you want to keep working on it, that'd be swell. But no worries if you don't have time. If anybody wants to drive it forward, please do post your own PR - and if you use this as a start, consider adding a co-author attribution at the end of your PR description. Thanks! 😊

**Pull request closed without merging**